### DISCUSSION TITLE
Setting DPDK with O-RU

### DISCUSSION THREAD

**Discussion by aibtw** [srsran/srsRAN_Project#518] [General Help] [answered]
Hi, 

Has anyone yet tried to connect DU to Benetel's RAN550 V0.8.1 with DPDK? 

As far as we understand, there is no way to do PTP timestamping on a network interface that is managed by DPDK (correct me if I am wrong). So, we prepared a setup to sync the DU and RU with 1PPS signals. 
Without DPDK, this setup worked, and we got to connect a UE to the network successfully. 

With this same setup, as soon as we bind the interface (between DU and RU) to vfio-pci, the UE doesn't connect anymore. The only changes we do are literally binding the network interface to vfio-pci and changing `network_interface` in DU configs, and adding `hal: eal_args: "--lcores (0-1)@(0-23) -a 0000:01:00.1"` to the configs as well. The gNB runs fine, and running `tssi` on the RU shows signal, but the UE can't connect. 

The gnb.log file gets filled with `[W] Missed incoming User-Plane PRACH messages for slot '941.18' and sector#0`. 

If anyone had a similar experience, would appreciate the help. 
Thank you.

**Comment by andrepuschmann** [comment]
Hey, if you use DPDK on one interface you can't run PTP on the same. To overcome this either use another port of the NIC to run the PTP daemon or use virtual functions for DPDK/gNB. You can then still run PTP on the physical interface.

**Reply by aibtw** [reply]
I see! Thank you for confirming. 
The RAN550 stated in their user manual that we shouldn't use their second SFP port. 
I tried virtual functions, but I wasn't able to get it right. I will definitely try it again. 

Will also try to share logs and pcaps as soon as I repeat the setup, if it may help.

**Reply by aibtw** [reply, marked as answer]
update: we were able to run the system with DPDK, using GPS sync mode instead of PTP, after firmware upgrade to v1.0.4.

Not sure if the firmware version is the only reason it worked, but I can't remember anything else that changed.

Same steps in srs guide were followed, in addition to Intel's E810 dpdk guide which was quite helpful: 
https://edc.intel.com/content/www/us/en/design/products/ethernet/config-guide-e810-dpdk/bios-settings/

**Reply by sushila-s** [reply]
> Hey, if you use DPDK on one interface you can't run PTP on the same. To overcome this either use another port of the NIC to run the PTP daemon or use virtual functions for DPDK/gNB. You can then still run PTP on the physical interface.  

@andrepuschmann We have the same problem of how to run PTP + DPDK. Do you have instructions on how to configure the IOMMU groups to do either of your suggestions above? Both don't work for us:   

1) We have one NIC with two ports and both are part of the same IOMMU group. So when we bind only one of the ports to DPDK, the DPDK bind works but when we run srsran, it complains.  

Our base set up: (enp4s0f0np0 and enp4s0f1np1 are the two ports on our NIC, we are trying to use enp4s0f0np0)
```    
0000:04:00.0 'Ethernet Controller E810-XXV for SFP 159b' if=enp4s0f0np0 drv=ice unused=vfio-pci *Active*  
0000:04:00.1 'Ethernet Controller E810-XXV for SFP 159b' if=enp4s0f1np1 drv=ice unused=vfio-pci *Active*.  
0000:08:00.0 'Ethernet Controller I225-V 15f3' if=eno1 drv=igc unused=vfio-pci *Active*. 
0000:09:00.0 'MT7921K (RZ608) Wi-Fi 6E 80MHz 0608' if=wlp9s0 drv=mt7921e unused=vfio-pci. 
```

srsran output:   
```
EAL: Detected CPU lcores: 24
EAL: Detected NUMA nodes: 1
EAL: Detected shared linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'VA'
EAL: VFIO support initialized
EAL: 0000:04:00.0 VFIO group is not viable! Not all devices in IOMMU group bound to VFIO or unbound
EAL: Requested device 0000:04:00.0 cannot be used
EAL: Bus (pci) probe failed.
TELEMETRY: No legacy callbacks, legacy socket not created
```

2) When we create a VF, this is what we have (new PCIE addresses and interface names)

```
Network devices using DPDK-compatible driver
============================================
0000:05:00.0 'Ethernet Adaptive Virtual Function 1889' drv=vfio-pci unused=iavf

Network devices using kernel driver
===================================
0000:04:00.0 'Ethernet Controller E810-XXV for SFP 159b' if=enp4s0f0np0 drv=ice unused=vfio-pci *Active*
0000:04:00.1 'Ethernet Controller E810-XXV for SFP 159b' if=enp4s0f1np1 drv=ice unused=vfio-pci *Active*
0000:08:00.0 'Ethernet Controller I225-V 15f3' if=eno1 drv=igc unused=vfio-pci *Active*
0000:09:00.0 'MT7921K (RZ608) Wi-Fi 6E 80MHz 0608' if=wlp9s0 drv=mt7921e unused=vfio-pci
```


We run PTP on PF enp4s0f0np0 and use 0000:05:00.0 in the srsran config file for DPDK  

srsrran complains again: 
```
EAL: Detected CPU lcores: 24
EAL: Detected NUMA nodes: 1
EAL: Detected shared linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'VA'
EAL: VFIO support initialized
EAL: 0000:05:00.0 VFIO group is not viable! Not all devices in IOMMU group bound to VFIO or unbound
EAL: Requested device 0000:05:00.0 cannot be used
EAL: Bus (pci) probe failed.
TELEMETRY: No legacy callbacks, legacy socket not created
```

Can you help us with the instructions on how to configure this correctly for our config?

**Reply by 00-day** [reply]
hi @sushila-s did you ever figure this out?

**Comment by debashish216** [comment]
@sushila-s, 
Just wondering if you managed to solve it. 
Thanks in advance to advise some work around pointers. :)
